src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationService.kt

```diff
@@ -0,0 +1,333 @@
+        }
+        application.support.add(createApplicationSupport(application, it.type, it.text))
+      }
+    }
+
+    val visit = this.visitRepo.findVisitByApplicationReference(applicationReference)
+    visit?.let {
+      application.reservedSlot = isReservationRequired(visit, application.sessionSlot, application.restriction)
+    }
+
+    val telemetryData = HashMap<String, String>()
+    telemetryData["applicationReference"] = application.reference
+    visit?.let {
+      telemetryData["bookingReference"] = it.reference
+    }
+    telemetryData["reservedSlot"] = application.reservedSlot.toString()
+
+    telemetryClientService.trackEvent(
+      VISIT_SLOT_CHANGED_EVENT,
+      telemetryData,
+    )
+
+    return applicationDtoBuilder.build(application)
+  }
+
+  private fun isReservationRequired(
+    visit: Visit,
+    newSessionSlot: SessionSlot,
+    newRestriction: VisitRestriction,
+  ): Boolean {
+    return visit.visitRestriction != newRestriction || visit.sessionSlotId != newSessionSlot.id
+  }
+
+  private fun isReservationRequired(
+    visit: Visit,
+    newSessionSlot: SessionSlot,
+    newRestriction: CreateApplicationRestriction,
+  ): Boolean {
+    return !newRestriction.isSame(visit.visitRestriction) || visit.sessionSlotId != newSessionSlot.id
+  }
+
+  @Transactional(readOnly = true)
+  fun findExpiredApplicationReferences(): List<String> {
+    LOG.debug("Entered findExpiredApplicationReferences : ${getExpiredApplicationDateAndTime()}")
+    return applicationRepo.findExpiredApplicationReferences(getExpiredApplicationDateAndTime())
+  }
+
+  fun getExpiredApplicationDateAndTime(): LocalDateTime {
+    return LocalDateTime.now().minusMinutes(expiredPeriodMinutes.toLong())
+  }
+
+  fun deleteAllExpiredApplications(applicationReferences: List<String>) {
+    applicationReferences.forEach {
+      val applicationToBeDeleted = getApplicationEntity(it)
+      val deleted = applicationRepo.deleteExpiredApplications(getExpiredApplicationDateAndTime(), it)
+      if (deleted > 0) {
+        val bookEvent = telemetryClientService.createApplicationTrackEventFromVisitEntity(applicationToBeDeleted)
+        telemetryClientService.trackEvent(VISIT_SLOT_RELEASED_EVENT, bookEvent)
+      }
+    }
+  }
+
+  fun isApplicationCompleted(reference: String): Boolean {
+    return applicationRepo.isApplicationCompleted(reference)
+  }
+
+  fun getApplicationEntity(applicationReference: String): Application {
+    return applicationRepo.findApplication(applicationReference)
+      ?: throw VisitNotFoundException("Application (reference $applicationReference) not found")
+  }
+
+  private fun saveEventAudit(
+    actionedBy: String,
+    application: ApplicationDto,
+    visit: Visit? = null,
+    applicationMethodType: ApplicationMethodType,
+  ) {
+    eventAuditRepository.saveAndFlush(
+      EventAudit(
+        actionedBy = actionedBy,
+        bookingReference = visit?.let { visit.reference } ?: "",
```

Comment:
null instead of ""?

Reply:
not sure could be null... it's probably better not to include it, I will do that